### ISSUE TITLE
Issues with ElggFile::detectMimeType() 

### ISSUE THREAD

**Issue by hypeJunction** [Elgg/Elgg#9010]
I am seeing strange behaviour ever since we introduced https://github.com/Elgg/Elgg/pull/8889.

When calling `ElggFile::detectMimeType()` from a contcrete class instance (other than ElggFile), `$this` is considered set and refers to the concrete instance. As a result the script is WSODing, as `getFilenameOnFilestore` does not exist for that concrete class. 

We should deprecate the static use, and add another static method. In the meantime, we should perhaps make sure that `$this` is an instance of `ElggFile`

**Comment by mrclay** [comment]
> When calling ElggFile::detectMimeType() from a contcrete class instance (other than ElggFile), $this is considered set

I don't think PHP does this. https://3v4l.org/oiHVg

**Comment by mrclay** [comment]
Wait, it does! I read the output wrong

**Comment by mrclay** [comment]
Your suggestion SGTM

**Comment by hypeJunction** [comment]
What are you thoughts on deprecation? Should we just convert to:

```php

public static function detectMimeType($file, $default = null) {
    if (!isset($file)) {
        // deprecation notice
    }
}
```

**Comment by mrclay** [comment]
Fixed by #9015
